crdstore/crdstore.go

```diff
@@ -0,0 +1,150 @@
+package crdstore
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"errors"
+
+	"sync"
+	"sync/atomic"
+	"unsafe"
+
+	"io/ioutil"
+
+	"github.com/allegro/akubra/crdstore/config"
+	"github.com/allegro/akubra/log"
+	"golang.org/x/sync/syncmap"
+)
+
+const (
+	keyPattern                   = "%s_____%s"
+	requestOptionsRequestTimeout = 100 * time.Millisecond
+	refreshTTLPercent            = 80 // Background refresh after refreshTTLPercent*TTL
+)
+
+// ErrCredentialsNotFound - Credential for given accessKey and backend haven't been found in yaml file
+var ErrCredentialsNotFound = errors.New("credentials not found")
+
+// CredentialStore instance
+var instances map[string]*CredentialsStore
+
+// CredentialsStore - gets a caches credentials from akubra-crdstore
+type CredentialsStore struct {
+	endpoint string
+	cache    *syncmap.Map
+	TTL      time.Duration
+	lock     sync.Mutex
+}
+
+// GetInstance - Get crdstore instance for endpoint
+func GetInstance(endpointName string) (instance *CredentialsStore, err error) {
+	if instance, ok := instances[endpointName]; ok {
+		return instance, nil
+	}
+	return nil, fmt.Errorf("error credentialStore `%s` is not defined", endpointName)
+}
+
+// InitializeCredentialsStore - Constructor for CredentialsStore
+func InitializeCredentialsStore(storeMap config.CredentialsStoreMap) {
+	instances = make(map[string]*CredentialsStore)
+	for name, cfg := range storeMap {
+		instances[name] = &CredentialsStore{
+			endpoint: cfg.Endpoint.String(),
+			cache:    new(syncmap.Map),
+			TTL:      cfg.CacheTTL.Duration,
+		}
+	}
+}
+
+func (cs *CredentialsStore) prepareKey(accessKey, backend string) string {
+	return fmt.Sprintf(keyPattern, accessKey, backend)
+}
+
+func (cs *CredentialsStore) updateCache(accessKey, backend, key string, csd *CredentialsStoreData, blocking bool) (newCsd *CredentialsStoreData, err error) {
+	if !blocking {
+		if !cs.tryLock() {
+			return csd, nil
+		}
+	} else {
+		cs.lock.Lock()
+	}
+	newCsd, err = cs.GetFromService(cs.endpoint, accessKey, backend)
+	switch {
+	case err == nil:
+		newCsd.err = nil
+	case err == ErrCredentialsNotFound:
+		newCsd = &CredentialsStoreData{EOL: time.Now().Add(cs.TTL), err: ErrCredentialsNotFound}
+	default:
+		if csd == nil {
+			newCsd = &CredentialsStoreData{EOL: time.Now().Add(cs.TTL), err: err}
+		} else {
+			*newCsd = *csd
+		}
+		newCsd.err = err
+		log.Printf("Error while updating cache for key `%s`: `%s`", key, err)
+	}
+	newCsd.EOL = time.Now().Add(cs.TTL)
+	cs.cache.Store(key, newCsd)
+	cs.lock.Unlock()
+	if newCsd.AccessKey == "" {
+		return nil, newCsd.err
+	}
+	return newCsd, nil
+}
+
+func (cs *CredentialsStore) tryLock() bool {
+	return atomic.CompareAndSwapInt32((*int32)(unsafe.Pointer(&cs.lock)), 0, 1)
```

Comment:
why do we need it? Add a comment

Reply:
For non blocking credentials update operations. Open question is If we need those. I doubt, so we'll clear things up before merge to master